File: Classes/SBBLKGUISearchField.m
//
//  SBBLKGUISearchField.m
//  Sunrise
//
//  Created by Atsushi Jike on 10/03/21.
//  Copyright 2010 Atsushi Jike. All rights reserved.
//

#import "SBBLKGUISearchField.h"
#import "SBUtil.h"

@implementation SBBLKGUISearchField

+ (void)initialize
{
	if (self == [SBBLKGUISearchField class])
	{
		[self setCellClass:[SBBLKGUISearchFieldCell class]];
	}
}

+ (Class)cellClass
{
	return [SBBLKGUISearchFieldCell class];
}

- (id)initWithFrame:(NSRect)rect
{
	if (self = [super initWithFrame:rect])
	{
		[self setDefaultValues];
	}
	return self;
}

- (void)setDefaultValues
{
	[self setAlignment:NSLeftTextAlignment];
	[self setDrawsBackground:NO];
	[self setTextColor:[NSColor whiteColor]];
}

@end

@implementation SBBLKGUISearchFieldCell

- (id)init
{
	if (self = [super init])
	{
		[self setDefaultValues];
	}
	return self;
}

- (void)setDefaultValues
{
	NSButtonCell *searchButtonCell = nil;
	searchButtonCell = [self searchButtonCell];
	[self setWraps:NO];
	[self setScrollable:YES];
	[self setFocusRingType:NSFocusRingTypeExterior];
	[searchButtonCell setImage:[NSImage imageNamed:@"Search.png"]];
	[searchButtonCell setAlternateImage:[NSImage imageNamed:@"Search.png"]];
}

- (NSText *)setUpFieldEditorAttributes:(NSText *)textObj
{
	NSText *text = [super setUpFieldEditorAttributes:textObj];
	if ([text isKindOfClass:[NSTextView class]])
	{
		NSDictionary *attributes = [NSDictionary dictionaryWithObjectsAndKeys:
									[NSColor whiteColor], 
									NSForegroundColorAttributeName, 
									[NSColor grayColor], 
									NSBackgroundColorAttributeName, nil];
		[(NSTextView *)text setInsertionPointColor:[NSColor whiteColor]];
		[(NSTextView *)text setSelectedTextAttributes:attributes];
	}
	return text;
}

- (void)drawWithFrame:(NSRect)cellFrame inView:(NSView *)controlView
{
	CGRect r = CGRectZero;
	CGContextRef ctx = [[NSGraphicsContext currentContext] graphicsPort];
	CGPathRef path = nil;
	CGFloat alpha = [controlView respondsToSelector:@selector(isEnabled)] ? ([(NSTextField *)controlView isEnabled] ? 1.0 : 0.2) : 1.0;
	
	r = NSRectToCGRect(cellFrame);
	path = SBRoundedPath(r, r.size.height / 2, 0, YES, YES);
	CGContextSaveGState(ctx);
	CGContextAddPath(ctx, path);
	CGContextSetRGBFillColor(ctx, 0.0, 0.0, 0.0, alpha * 0.1);
	CGContextFillPath(ctx);
	CGContextRestoreGState(ctx);
	CGPathRelease(path);
	
	r = NSRectToCGRect(cellFrame);
	r.origin.x += 0.5;
	r.origin.y += 0.5;
	r.size.width -= 1.0;
	r.size.height -= 1.0;
	path = SBRoundedPath(r, r.size.height / 2, 0, YES, YES);
	CGContextSaveGState(ctx);
	CGContextAddPath(ctx, path);
	CGContextSetLineWidth(ctx, 0.5);
	CGContextSetRGBStrokeColor(ctx, 1.0, 1.0, 1.0, alpha);
	CGContextStrokePath(ctx);
	CGContextRestoreGState(ctx);
	CGPathRelease(path);
	
	[self drawInteriorWithFrame:cellFrame inView:controlView];
}

- (void)drawInteriorWithFrame:(NSRect)cellFrame inView:(NSView *)controlView
{
	[super drawInteriorWithFrame:cellFrame inView:controlView];
}

- (void)editWithFrame:(NSRect)aRect inView:(NSView *)controlView editor:(NSText *)textObj delegate:(id)anObject event:(NSEvent *)theEvent
{
	[super editWithFrame:aRect inView:controlView editor:textObj delegate:anObject event:theEvent];
}

- (void)selectWithFrame:(NSRect)aRect inView:(NSView *)controlView editor:(NSText *)textObj delegate:(id)anObject start:(int)selStart length:(int)selLength
{
	[super selectWithFrame:aRect inView:controlView editor:textObj delegate:anObject start:selStart length:selLength];
}

@end

File: Classes/SBBLKGUISearchField.swift
/*
SBBLKGUISearchField.swift

Copyright (c) 2014, Alice Atlas
Copyright (c) 2010, Atsushi Jike
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

class SBBLKGUISearchField: NSSearchField {
    override class func initialize() {
        SBBLKGUISearchField.setCellClass(SBBLKGUISearchFieldCell.self)
    }
    
    override init(frame: NSRect) {
        super.init(frame: frame)
        setDefaultValues()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    private func setDefaultValues() {
        alignment = .LeftTextAlignment
        drawsBackground = false
        textColor = NSColor.whiteColor()
    }
}

class SBBLKGUISearchFieldCell: NSSearchFieldCell {
    override init() {
        super.init()
    }
    
    @objc(initTextCell:)
    override init(textCell string: String) {
        super.init(textCell: string)
        setDefaultValues()
    }
    
    required init(coder: NSCoder) {
        super.init(coder: coder)
    }

    private func setDefaultValues() {
        wraps = false
        scrollable = true
        focusRingType = .Exterior
        searchButtonCell!.image = NSImage(named: "Search.png")!
        searchButtonCell!.alternateImage = searchButtonCell!.image

    }
    
    override func setUpFieldEditorAttributes(textObj: NSText) -> NSText {
        let text = super.setUpFieldEditorAttributes(textObj)
        if let textView = text as? NSTextView {
            let attributes = [NSForegroundColorAttributeName: NSColor.whiteColor(),
                              NSBackgroundColorAttributeName: NSColor.grayColor()]
            textView.insertionPointColor = NSColor.whiteColor()
            textView.selectedTextAttributes = attributes
        }
        return text
    }
    
    override func drawWithFrame(cellFrame: NSRect, inView: NSView) {
        let controlView = inView as? NSControl
        let alpha: CGFloat = (controlView?.enabled ?? true) ? 1.0 : 0.2
        var r = cellFrame
        var radius = r.size.height / 2
        var path = NSBezierPath(roundedRect: r, xRadius: radius, yRadius: radius)
        NSColor(deviceWhite: 0.0, alpha: alpha * 0.1).set()
        path.fill()
        
        r.inset(dx: 0.5, dy: 0.5)
        radius = r.size.height / 2
        path = NSBezierPath(roundedRect: r, xRadius: radius, yRadius: radius)
        path.lineWidth = 0.5
        NSColor(deviceWhite: 1.0, alpha: alpha).set()
        path.stroke()
        
        drawInteriorWithFrame(cellFrame, inView: inView)
    }
}